packages/opentelemetry-plugin-express/src/express.ts

```diff
@@ -204,12 +204,37 @@ export class ExpressPlugin extends BasePlugin<typeof express> {
             return patchEnd(span, plugin._tracer.bind(next))();
           };
         }
+        /**
+         * Ending a layer's span at the right time is quite difficult because
+         * of their nature: they can be either sync or async and either it call
+         * their callback or not, and we can't know that before calling it.
+         *
+         * Here are the different state that can result from calling the layer code:
+         *  1. The layer is sync and call its callback
+         *      - it will be correctly closed by the above code
+         *  2. The layer is sync and never call its callback
+         *      - we need to end the span manually after calling the layer code
+         *  3. The layer is async and call its callback
+         *      - it will be correctly closed by the above code
+         *  4. The layer is async and never call its callback
+         *      - we need to end the span manually
+         *      - however it will not represent the actual time that the span took
+         *      - we would need to hook into the http response's `end` event.
+         *
+         * The fix for case 4) is to hook into the `end` event, but it would apply
+         * to case 2) and 3), which would cost quite a lot of memory for big express apps.
+         *
+         * The workaround is to schedule a callback to run on next tick and check if the
```

Comment:
Doesn't this break case 3? If we end the span immediately, we are not waiting for the callback to be called. Unless I am misunderstanding the above code?

Reply:
You are right but i don't see how we can do better

Reply:
In my opinion case (3) is far more important and common than (4). I would be against optimizing for (4) with any method that breaks (3). Other solutions I can think of:

1. end all spans when response ends (`res.on('end')` or similar) - I do not like this solution
2. end spans after some max wait time
```typescript
const maxWaitTimer = setTimeout(() => {
	endAllSpans()
}, maxWaitTime);

// if callback is called
clearTimeout(maxWaitTimer);
```
3. end _all_ spans that don't end in some max time
```typescript
class Span {
    private _ended = false;
    private _maxWaitTimer: NodeJS.Timeout;

    constructor() {
        this._maxWaitTimer = setTimeout(() => {
            if (!this._ended) {
				// probably log here
                this.end();
            }
        })
    }

    public end() {
        if (this._ended) return;
        clearTimeout(this._maxWaitTimer);
        this._ended = true;
    }
}
```

Reply:
I know census closed child spans that weren't ended when the root span was ended but i'm pretty sure that would not be doable with our current API, right ?

All solutions seems to be wrong in term of time spent anyway, could we just acknowledge that and just end them in all cases directly ?

Reply:
Why would that not be doable with our API?

Reply:
Maybe i didn't find it but i believe we store no reference of spans created from a given root span (since there is no such concept) ? That would be the only way to end them after the end of the http response

Reply:
I went with listening for the `end` event. I believe we should aim for correctness and following feedback from users maybe add an option to control this behavior. WDYT ?

Reply:
Actually the implementation wasn't working, i ended up with ended the span directly and documenting that behavior. We could open up an issue to fix that on separate PR if you have an idea on how to do it properly

Reply:
@mayurkale22 wdyt about this?

Reply:
@mayurkale22 friendly ping :)

Reply:
> I know census closed child spans that weren't ended when the root span was ended but i'm pretty sure that would not be doable with our current API, right ?

In census, we keep references to all child spans and that's why it is possible to close them when the root span was ended. But this is not the case in OpenTelemetry implementation.

Reply:
> Actually the implementation wasn't working, i ended up with ended the span directly and documenting that behavior. We could open up an issue to fix that on separate PR if you have an idea on how to do it properly

sounds ok for now.